wasm.md

```diff
@@ -589,37 +590,73 @@ The `*_local` instructions are defined here.
  // --------------------------------------------------------------
 
-    syntax Instr ::= "(" "func"              FuncDecls Instrs ")"
-                   | "(" "func" Identifier FuncDecls Instrs ")"
-                   | "func" Identifier "::" FuncType VecType "{" Instrs "}"
- // -----------------------------------------------------------------------
-    rule <k> ( func FDECLS INSTRS )
-          => func gatherExportedName(FDECLS) :: gatherFuncType(FDECLS) gatherTypes(local, FDECLS) { INSTRS }
-         ...
+    syntax FuncType ::= gatherFuncType ( FuncDecls ) [function]
+ // -----------------------------------------------------------
+    rule gatherFuncType(FDECLS) => gatherTypes(param, FDECLS) -> gatherTypes(result, FDECLS)
+
+    syntax VecType ::=  gatherTypes ( TypeKeyWord , FuncDecls )            [function]
+                     | #gatherTypes ( TypeKeyWord , FuncDecls , ValTypes ) [function]
+ // ---------------------------------------------------------------------------------
+    rule gatherTypes(TKW, FDECLS) => #gatherTypes(TKW, FDECLS, .ValTypes)
+
+    rule #gatherTypes(TKW , .FuncDecls            , TYPES) => [ TYPES ]
+    rule #gatherTypes(TKW , FDECL:FuncDecl FDECLS , TYPES) => #gatherTypes(TKW, FDECLS, TYPES) [owise]
+
+    rule #gatherTypes(TKW , TKW VTYPES' FDECLS:FuncDecls , VTYPES)
+      => #gatherTypes(TKW ,             FDECLS           , VTYPES + VTYPES')
+```
+
+### Function Export Definition
+
+```k
+    syntax FuncExport  ::= "(" "export" String ")"
+    syntax FuncExports ::= List{FuncExport, ""} [klabel(listFuncExport)] 
+    syntax Instr ::= #handleExports ( FuncExports )
+ // -----------------------------------------------
+    rule <k> #handleExports ( .FuncExports ) => . ... </k>
+    rule <k> #handleExports ( ( export ENAME ) FEXPO:FuncExports )
+          => ( export ENAME ( func NEXTIDX ) )
+          ~> #handleExports( FEXPO )
+          ...
          </k>
+         <nextFuncIdx> NEXTIDX </nextFuncIdx>
+```
 
-    rule <k> ( func FNAME FDECLS INSTRS )
-          => func FNAME :: gatherFuncType(FDECLS) gatherTypes(local, FDECLS) { INSTRS }
+### Function Declaration
+
+Function declarations can look quite different depending on which fields are ommitted and what the context is.
+Here, we allow for an "abstract" function declaration using syntax `func_::___`, and a more concrete one which allows arbitrary order of declaration of parameters, locals, and results.
+
+```k
+    syntax Instr ::= "(" "func"            FuncExports FuncDecls   Instrs ")"
+                   | "(" "func" Identifier FuncExports FuncDecls   Instrs ")"
+                   |     "func"            FuncType    VecType "{" Instrs "}"
+ // -------------------------------------------------------------------------
+
+    rule <k> ( func FEXPO:FuncExports FDECLS:FuncDecls INSTRS:Instrs )
+          => #handleExports( FEXPO ) ~> func gatherFuncType(FDECLS) gatherTypes(local, FDECLS) { INSTRS }
          ...
          </k>
```

Comment:
Can we do:

```
syntax Identifier ::= #asId ( Int )

rule <k> ( func ( export EXPNAME  ) FEXPS:FuncExports FDECLS:FuncDecls INSTRS:Instrs )
       => ( export EXPNAME ( func #asId(NEXTID) ) )
       ~> ( func #asId(NEXT) FEXPS FDECLS INSTRS )
      ...
    </k> <nextFuncId> NEXT => NEXT +Int 1 </nextFuncId>
```

Reply:
And a second rule:

```
rule <k> ( func FID ( export EXPNAME  ) FEXPS:FuncExports FDECLS:FuncDecls INSTRS:Instrs )
       => ( export EXPNAME ( func FID ) )
       ~> ( func FID FEXPS FDECLS INSTRS )
      ...
    </k>
```

Reply:
this can't work because we cannot increment <nextFuncId> before we bind it to an address of the store. Will fix it in another way.

Reply:
I don't see why this can't be the way it's handled. It's semantically equivalent and avoids having to use `#handleExports` at all.